File: 543.diameter-of-binary-tree.cpp
/*
 * @lc app=leetcode id=543 lang=cpp
 *
 * [543] Diameter of Binary Tree
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution
{
public:
    int ans = 0;
    int diameterOfBinaryTree(TreeNode *root)
    {
        height(root);
        return ans;
    }
    int height(TreeNode *node)
    {
        if (node==NULL)
            return 0;
        int l=height(node->right);
        int r=height(node->left);
        ans = max( l+r, ans);
        return 1 + max(l,r);
    }
};
// @lc code=end

File: 543.diameter-of-binary-tree.py
#
# @lc app=leetcode id=543 lang=python3
#
# [543] Diameter of Binary Tree
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    ans=0
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        self.height(root)

        return self.ans

    def height(self, node:Optional[TreeNode]) -> int:
        if not node:
            return 0
        l=self.height(node.right)
        r=self.height(node.left)
        self.ans = max( l+r, self.ans)
        return 1 + max(l,r)
# @lc code=end

